consensus/babe/verify.go

```diff
@@ -18,3 +20,76 @@ func (b *Session) verifySlotWinner(slot uint64, header *BabeHeader) (bool, error
 
 	return pub.VrfVerify(vrfInput, header.VrfOutput[:], header.VrfProof[:])
 }
+
+func (b *Session) verifyAuthorshipRight(slot uint64, header *types.BlockHeader) (bool, error) {
+	// header should have 2 digest items (possibly more in the future)
+	// first item should be pre-digest, second should be seal
+	if len(header.Digest) < 2 {
+		return false, fmt.Errorf("block header is missing digest items")
+	}
+
+	// check for valid seal by verifying signature
+	preDigestBytes := header.Digest[0]
+	sealBytes := header.Digest[len(header.Digest)-1]
+
+	digestItem, err := types.DecodeDigestItem(preDigestBytes)
+	if err != nil {
+		return false, err
+	}
+
+	preDigest, ok := digestItem.(*types.PreRuntimeDigest)
+	if !ok {
+		return false, fmt.Errorf("first digest item is not pre-digest")
+	}
+
+	digestItem, err = types.DecodeDigestItem(sealBytes)
+	if err != nil {
+		return false, err
+	}
+
+	seal, ok := digestItem.(*types.SealDigest)
+	if !ok {
+		return false, fmt.Errorf("last digest item is not seal")
+	}
+
+	babeHeader := new(BabeHeader)
+	err = babeHeader.Decode(preDigest.Data)
+	if err != nil {
+		return false, fmt.Errorf("cannot decode babe header from pre-digest: %s", err)
+	}
+
+	if len(b.authorityData) <= int(babeHeader.BlockProducerIndex) {
```

Comment:
If the length of authority data is less than or equal to the block producer index. I am a bit confused by the use of `int(babeHeader.BlockProducerIndex)` here. Maybe a comment to explain?

Reply:
the BlockProducerIndex identifies which authority in the authorityData list created the block. if the BlockProducerIndex is out of the range of the authorityData list (perhaps due to an invalid block), there will be a much worse error :p the code will panic